### PR TITLE
Add OnProcessStart callback to defer shutdown when not ready

### LINKED ISSUE #418: Race between Shutdown and starting a new process
## Bug report

**Required Info:**

- Operating System:
  - Fedora 32
- Installation type:
  - from source
- Version or commit hash:
  - cac43b1
- DDS implementation:
  - CycloneDDS
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue
```python
import launch

def generate_launch_description():
    return launch.LaunchDescription([
        launch.actions.ExecuteProcess(cmd=[
            launch.substitutions.FindExecutable(name='sleep'), '10'
        ]),
        launch.actions.Shutdown(),
    ])
```

#### Expected behavior
Launch starts the sleep process and then immediately sends Ctrl+C to that process, then exits after a very short amount of time.

#### Actual behavior
Launch waits 5 seconds to SIGTERM the process and prints the following error:
<details>

```
Task exception was never retrieved
future: <Task finished name='Task-8' coro=<LaunchService._process_one_event() done, defined at launch/launch/launch_service.py:271> exception=RuntimeError('Signal event received before subprocess transport available.')>
Traceback (most recent call last):
  File "launch/launch/launch_service.py", line 273, in _process_one_event
    await self.__process_event(next_event)
  File "launch/launch/launch_service.py", line 293, in __process_event
    visit_all_entities_and_collect_futures(entity, self.__context))
  File "launch/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 38, in visit_all_entities_and_collect_futures
    sub_entities = entity.visit(context)
  File "launch/launch/action.py", line 108, in visit
    return self.execute(context)
  File "launch/launch/actions/opaque_function.py", line 75, in execute
    return self.__function(context, *self.__args, **self.__kwargs)
  File "launch/launch/actions/execute_process.py", line 421, in __on_signal_process_event
    raise RuntimeError('Signal event received before subprocess transport available.')
RuntimeError: Signal event received before subprocess transport available.
```

</details>

#### Additional information

This is a contrived example, but it demonstrates the race between a launch shutdown and a process being spawned. I imagine that you could get it to happen if you Ctrl-C at just the right time during a process launch.

Another way to trigger this is to create a launch test that runs the tests REALLY FAST (like just running `assert True` or something). Since even a simple test like that takes a little time, launch fails sometimes and succeeds others.

Maybe a ShutdownProcess event arrives before the process has even started? I'm not sure.

### PR DESCRIPTION
**Pull request by Blast545**
Should close #418 and [demos#438](https://github.com/ros2/demos/issues/438).

The problem is generated when a shutdown condition reaches a process started by launch, that doesn't have its transport available. The exception is generated in the `execute_process.py` file:
https://github.com/ros2/launch/blob/3a6bde714a402121aeb64147ac3b2c8a172af79c/launch/launch/actions/execute_process.py#L421

As the processes are not available to receive the `shutdown` event, my proposed solution is to add a handler to call the shutdown procedure when the process(es) start executing. 

The exceptions generated there are not catched by the launch system, which might require further addressing.

Signed-off-by: Jorge Perez <jjperez@ekumenlabs.com>

### CODE REVIEW

**Comment by Blast545** [comment]
CI:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10740/)

File: launch/launch/actions/execute_process.py [hunk 1]
```diff
@@ -378,10 +379,19 @@ def get_sub_entities(self):
             return self.__on_exit
         return []
 
+    def __deferred_shutdown(self, event, context):
+        return self._shutdown_process(context, send_sigint=True)
```
Comment (hidmic) [inline launch/launch/actions/execute_process.py:383]: @Blast545 I'd think `send_sigint` has to have the same value it had when `_shutdown_process()` was first called.

File: launch/launch/actions/execute_process.py [hunk 2]
```diff
@@ -378,10 +379,19 @@ def get_sub_entities(self):
...
+
     def _shutdown_process(self, context, *, send_sigint):
         if self.__shutdown_received:
             # Do not handle shutdown more than once.
             return None
+
+        if self.process_details is None or self._subprocess_transport is None:
+            # Skip shutting down if the process is not ready
+            context.register_event_handler(OnProcessStart(on_start=self.__deferred_shutdown))
```
Comment (hidmic) [inline launch/launch/actions/execute_process.py:392]: @Blast545 how do you prevent this event handler being added multiple times?

Comment (Blast545) [reply launch/launch/actions/execute_process.py:392]: @hidmic The handler itself might be added multiple times if the shutdown process is called multiple times before the subprocess_transport is available.

However, the first check of the function `if self.__shutdown_received:` doesn't allow for the shutdown actions to be added to the actions queue multiple times.

File: launch/launch/actions/execute_process.py (hunk 1 above)
Comment (Blast545) [reply launch/launch/actions/execute_process.py:383]: I agree, I'll add a commit to address this.

**Comment by Blast545** [comment]
CI:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10761/)

File: launch/launch/actions/execute_process.py [hunk 3]
```diff
@@ -382,6 +383,15 @@ def _shutdown_process(self, context, *, send_sigint):
...
+        if self.process_details is None or self._subprocess_transport is None:
+            # Skip shutting down if the process is not ready
+            context.register_event_handler(
+                OnProcessStart(
+                    on_start=lambda event, context:
+                    self._shutdown_process(context, send_sigint=send_sigint)))
+            return None
+
         self.__shutdown_received = True
         if self.__completed_future is None:
```
Comment (cottsay) [inline launch/launch/actions/execute_process.py:396]: How is this conditional related to the one you're adding? From the comment:
> Execution not started so nothing to do

Is the trace in #418 appearing because the process is in an intermediate state between "not started" and "running"? Should the new conditional should be processed after this one?

Comment (Blast545) [reply launch/launch/actions/execute_process.py:396]: The trace in #418 appears because the process haven't started before the shutdown process and this attempts to send it a `SIGINT` signal.

With this new conditional the shutdown signal is not sent to processes not started or not available yet and sent after the events have started.

Although I've discussed this fix with @ivanpauno and he told me the correct approach would be making the shutdown process to check if a process haven't started and never start it all. But I didn't find and easy way to implement that.

File: launch/launch/actions/execute_process.py [hunk 4]
```diff
@@ -382,6 +383,15 @@ def _shutdown_process(self, context, *, send_sigint):
         if self.__shutdown_received:
             # Do not handle shutdown more than once.
             return None
+
+        if self.process_details is None or self._subprocess_transport is None:
+            # Skip shutting down if the process is not ready
+            context.register_event_handler(
+                OnProcessStart(
```
Comment (ivanpauno) [inline launch/launch/actions/execute_process.py:390]: This doesn't seem to be the correct approach to me.
In this case, we don't need to do anything. Only setting `self.__shutdown_received` to `True` and returning.

**Comment by sloretz** [comment]
@Blast545 friendly ping. What do you think of @ivanpauno's feedback?

**Comment by Blast545** [comment]
@sloretz I do think his thoughts on the problem would lead to a better solution but I've tried his proposed comments without success. I'd probably take some time to check again with him early next week.

**Comment by ivanpauno** [comment]
Closing as https://github.com/ros2/launch/pull/427 fixed this https://github.com/ros2/launch/pull/418.
